app/hooks/usePagination.ts

```diff
@@ -0,0 +1,74 @@
+import { useMemo, useState } from "react"
+import { IdObj, PaginateProps } from "~/types/table"
+
+type UsePaginationProps = {
+  data: IdObj[]
+  paginate: boolean | PaginateProps
+  search: boolean
+  searchTerm: string
+}
+
+type UsePaginationReturn = {
+  paginatedData: IdObj[]
+  perPage: number
+  totalPages: number
+  totalData: IdObj[]
+  page: number
+  handlePageChange: (newPage: number) => void
+}
+
+export const usePagination = ({
+  data,
+  paginate,
+  search,
+  searchTerm,
+}: UsePaginationProps): UsePaginationReturn => {
+  const [page, setPage] = useState(
+    typeof paginate === "boolean" || !paginate ? 1 : paginate.page ?? 1,
+  )
+
+  const perPage = useMemo(
+    () => (typeof paginate === "boolean" || !paginate ? 5 : paginate.perPage ?? 5),
+    [paginate],
+  )
+
+  const totalData = useMemo(() => {
```

Comment:
Since we're at it... I'd love to modify and improve this code.

```
  const totalData = useMemo(() => {
    if (search && searchTerm) {
      return data.filter((row) => {
        const columns = Object.values(row).map((column) =>
          typeof column === "object" ? column.value : column,
        )
        const filter = columns.join().toLowerCase().trim()
        const exists = columns.some((column) =>
          column.toLowerCase().includes(searchTerm.toLowerCase().trim()),
        )
        return exists
      })
    }
    return data
  }, [data, search, searchTerm])
```
- Avoid using let and directly return convenient data 
- Avoid using if statements under map and use inline ternary operator
- Use short version of return under map function (no brackets)

Reply:
great observations @RabeeAbuBaker, I resolved the suggestions and also improved the `paginatedData` below, let me know if anything

Reply:
Great changes @r1ck12 👏